### PR TITLE
Added System Management Vulnerability

### LINKED ISSUE #647: Adding Session Authentication related vulnerabilities
VulnerableApp currently has Authentication related vulnerabilites but it never had session related vulnerabilities.

The goal of this vulnerabilities is to introduce sessions to VulnerableApp.

**Session management**
1. Session fixation (CWE-384) -- Session ID not regenerated after login. Attacker pre-sets the session cookie.
2. Predictable session IDs (CWE-330) -- Session tokens generated from a sequential counter instead of a CSPRNG.
3. Missing logout invalidation (CWE-613) -- Logout clears the cookie but does not destroy the server-side session.
4. No login rate limiting (CWE-307) -- Unlimited login attempts with no delay, lockout, or CAPTCHA.

**Password reset**
5. Predictable reset token -- Token generated from timestamp or sequential ID.
6. Reset token reuse -- Token not invalidated after first use.
7. Reset without identity verification -- No email/security question check before allowing password change.

We are planning to add local SMTP server as well, if this issue is picked after that, please try adding scenarios of password reset with emails.

### PR DESCRIPTION
**Pull request by CHI7440**
Resolve #647 

This PR supersedes #653.

While syncing my fork with the upstream repository, I accidentally used GitHub's **Sync fork → Discard commits** option, which reset my fork's `master` branch and caused GitHub to automatically close PR #653.

I recovered the original commits locally, rebased them on the latest upstream changes, and recreated this PR.

Current Status

✅ All previously implemented changes have been preserved.
✅ Several review comments from #653 have already been addressed.
⏳ A few review threads are intentionally left pending because they require clarification before implementation.

I apologize for the inconvenience and appreciate your time reviewing this contribution again.

Thank you!

## Summary by CodeRabbit

* **New Features**
  * Added session-management training with six interactive security levels, each providing level-specific login, profile, and logout behavior (including deterministic vs random session IDs, optional invalidation, and failed-attempt blocking/rate-limiting indicators).
  * Introduced the session-management web UI, including new page templates, styling, client-side login/profile/logout flow, and i18n challenge content with example payloads.
  * Extended the vulnerability catalog with new session-management attack vectors and updated the related database schema/seed data.
* **Tests**
  * Added unit tests validating login/profile/logout semantics and failure/edge-case behavior across all levels.

### CODE REVIEW

File: src/main/resources/i18n/messages.properties [hunk 1]
```diff
@@ -472,6 +472,51 @@ AUTHENTICATION_VULNERABILITY_LEVEL_10_HINT_3=Because the cost is only 4, a commo
...
+PREDICTABLE_SESSION_ID_DESCRIPTION=The application generates session IDs using a predictable counter instead of secure random values. Attackers can guess nearby session tokens and access active sessions.
+OBSCURED_PREDICTABLE_SESSION_ID_DESCRIPTION=The application generates session IDs using a predictable counter and a weak encoding methods. Attackers can predict or brute-force session tokens to access active sessions.
+MISSING_LOGOUT_INVALIDATION_DESCRIPTION=The application clears the browser cookie during logout but does not destroy the server-side session. Reusing the old session token still allows access to the authenticated profile.
+NO_LOGIN_RATE_LIMITING_DESCRIPTION=The application allows repeated login attempts without delay, lockout, throttling, or CAPTCHA. Attackers can keep guessing credentials without restriction.
+
+#Challenge Cards for Session Management Vulnerabilities
+SESSION_MANAGEMENT_LEVEL1_CHALLENGE=Log in with a session token you supply before authentication. The application does not regenerate the session token after login.
```
Comment (preetkaran20) [inline src/main/resources/i18n/messages.properties]: in PR https://github.com/SasanLabs/VulnerableApp/pull/653 i have added comments on what does challenge mean. can you please fix suggestions like https://github.com/SasanLabs/VulnerableApp/pull/653#discussion_r3470930341
https://github.com/SasanLabs/VulnerableApp/pull/653#discussion_r3470937612
https://github.com/SasanLabs/VulnerableApp/pull/653#discussion_r3470994591

Comment (preetkaran20) [reply src/main/resources/i18n/messages.properties]: https://github.com/SasanLabs/VulnerableApp/pull/653#discussion_r3470900131 this one too.

**Review by preetkaran20 — CHANGES_REQUESTED** [review]

**Comment by preetkaran20** [comment]
image

I can still see same issues. can you please address them?

File: src/main/resources/i18n/messages.properties (hunk 1 above)
Comment (CHI7440) [reply src/main/resources/i18n/messages.properties]: Apologies for the delay, I was thinking of the way and before I start changing the PR, I want to confirm if I understood the direction correctly.

For the UI, I am thinking we should make it behave like a normal application. So the user logs in, and the profile is shown automatically. On logout, the profile is hidden. I will remove the explicit “Check Profile” button, token input, and raw JSON response from the page.

I will still keep the backend profile endpoint, because that is what the learner/attacker will try to abuse manually using browser dev tools, curl, Postman, etc. So the UI will look normal, but the vulnerability can still be discovered by replaying requests or manipulating cookies/session tokens.

For the challenge cards, I will rewrite them more like attacker goals instead of direct instructions. For example, “can you hijack the victim’s session without knowing their password?” instead of “log in and observe the token.”

For users, since this PR is focused on session management and not password vulnerabilities, I am thinking of keeping it simple: one attacker user and one or more victim users shared across the session levels. Level 5 can still have a victim password that makes sense for demonstrating no login rate limiting.

Also, I will keep password reset vulnerabilities out of this PR and leave them for the separate ticket.

Does this direction sound okay before I start making the changes?

Thanks for the patience!!

Comment (preetkaran20) [reply src/main/resources/i18n/messages.properties]: On UI point: Agreed
Keeping backend profile endpoint: Agreed and great point. UI will call the profile Api on successful login?
Challenge cards: Yes Agreed
Agreed on point that it is session management. perfect !!!
Yes, please create a seperate ticket for reset vunerability.

Agreed on all points @CHI7440 Thanks for putting them here. 🚀

Comment (CHI7440) [reply src/main/resources/i18n/messages.properties]: @preetkaran20, done the changes having following points: 
- Expanded session management description
- Reworked challenge cards and hints into attacker-goal style
- Updated attack vector payload text to match cookie names/API replay flow
- Added attacker/victim/admin/rate-limit-victim session users
- Simplified UI to show profile automatically after login
- Removed visible token textbox, Check Profile button, and raw JSON response
- Updated session management tests for new default user

Please have a look into it, thanks!!

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                sessionUserRepository.findByUsernameAndPassword(username, password);
+        if (user.isEmpty()) {
+            return response("Invalid credentials", false);
+        }
+
+        String sessionId = "SESSION-" + level2PredictableSessionCounter.incrementAndGet();
+        sessions.put(sessionKey(LevelConstants.LEVEL_2, sessionId), user.get());
+
+        Map<String, Object> content = new LinkedHashMap<>();
+        content.put("message", SUCCESSFUL_LOGIN_MESSAGE);
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java]: can you please create string constants for message, sessionId etc. it is used in all the levels.

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+    public ResponseEntity<GenericVulnerabilityResponseBean<Object>> level2Login(
+            String username, String password) {
+        Optional<SessionUser> user =
+                sessionUserRepository.findByUsernameAndPassword(username, password);
+        if (user.isEmpty()) {
+            return response("Invalid credentials", false);
+        }
+
+        String sessionId = "SESSION-" + level2PredictableSessionCounter.incrementAndGet();
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java]: can you please create string constant for SESSION- too?

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+    public ResponseEntity<GenericVulnerabilityResponseBean<Object>> profile(
+            String level, String sessionId) {
+        if (sessionId == null
+                || sessionId.isBlank()
+                || !sessions.containsKey(sessionKey(level, sessionId))) {
+            return response("No valid session found", false);
+        }
+        SessionUser user = sessions.get(sessionKey(level, sessionId));
+        Map<String, Object> content = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java:252]: @CHI7440 can you please fix race conditions?

File: src/main/resources/static/templates/SessionManagementVulnerability/LEVEL_6/SecureSessionManagement.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  clearSessionProfile();
+  setSessionStatus(data.isValid ? "Logged out" : data.content);
+}
+
+document
+  .getElementById("sessionLoginBtn")
+  .addEventListener("click", function () {
+    var username = document.getElementById("sessionUsername").value;
+    var password = document.getElementById("sessionPassword").value;
+    var url =
```
Comment (preetkaran20) [inline src/main/resources/static/templates/SessionManagementVulnerability/LEVEL_6/SecureSessionManagement.js]: please pass username and password in request body as URL gets logged in access logs and it becomes security risk with username password logging.

File: src/main/resources/static/templates/SessionManagementVulnerability/Common/SessionManagement.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  document.getElementById("sessionProfileCard").style.display = "";
+  setSessionStatus("");
+}
+
+function loadSessionProfile() {
+  doGetAjaxCall(
+    showSessionProfile,
+    getUrlForVulnerabilityLevel() +
+      "?action=profile&token=" +
+      encodeURIComponent(currentSessionToken),
```
Comment (preetkaran20) [inline src/main/resources/static/templates/SessionManagementVulnerability/Common/SessionManagement.js]: we should not pass sessiontoken in URL, neither UI should touch session id. Cookies will be automatically passed from front end to backend and we should rely on that mechanism only. can you please change this?

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                @ChallengeCard.Hint(order = 2, text = "SESSION_MANAGEMENT_LEVEL1_HINT_2"),
+                @ChallengeCard.Hint(order = 3, text = "SESSION_MANAGEMENT_LEVEL1_HINT_3")
+            },
+            payload =
+                    @ChallengeCard.Payload(
+                            description = "SESSION_MANAGEMENT_LEVEL1_PAYLOAD_DESCRIPTION",
+                            value = "SESSION_MANAGEMENT_LEVEL1_PAYLOAD_VALUE"))
+    public ResponseEntity<GenericVulnerabilityResponseBean<Object>> level1SessionFixation(
+            @RequestParam(required = false) String action,
+            @RequestParam(required = false) String username,
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: can you please make them as part of Request body?

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            },
+            payload =
+                    @ChallengeCard.Payload(
+                            description = "SESSION_MANAGEMENT_LEVEL2_PAYLOAD_DESCRIPTION",
+                            value = "SESSION_MANAGEMENT_LEVEL2_PAYLOAD_VALUE"))
+    public ResponseEntity<GenericVulnerabilityResponseBean<Object>> level2PredictableSessionId(
+            @RequestParam(required = false) String action,
+            @RequestParam(required = false) String username,
+            @RequestParam(required = false) String password,
+            @RequestParam(required = false) String token) {
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: please use cookies instead of Request params for tokens

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    private final JdbcTemplate jdbcTemplate;
+
+    public SessionUserRepository(@Qualifier("applicationJdbcTemplate") JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public Optional<SessionUser> findByUsernameAndPassword(String username, String password) {
+        List<SessionUser> users =
+                jdbcTemplate.query(
+                        "SELECT * FROM session_users WHERE username = ? AND password = ?",
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java]: please use spring data respository (we just need to define an interface method) instead of manually querying using jdbc templates.

File: src/main/resources/i18n/messages.properties [hunk 10]
```diff
@@ -472,6 +472,52 @@ AUTHENTICATION_VULNERABILITY_LEVEL_10_HINT_3=Because the cost is only 4, a commo
...
+
+#Challenge Cards for Session Management Vulnerabilities
+SESSION_MANAGEMENT_LEVEL1_CHALLENGE=Hijack the victim's session without knowing their password. Pre-set a session cookie in the victim's browser and prove the application keeps using that same token after login.
+SESSION_MANAGEMENT_LEVEL1_HINT_1=In session fixation, the attacker chooses or plants a session token before authentication.
+SESSION_MANAGEMENT_LEVEL1_HINT_2=Set the VAPP_SESSION_LEVEL1 cookie to a value you control before the victim logs in.
+SESSION_MANAGEMENT_LEVEL1_HINT_3=After the victim logs in, manually call the profile API using curl, Postman, or browser dev tools with the same cookie value and check whether the victim profile is returned.
+SESSION_MANAGEMENT_LEVEL1_PAYLOAD_DESCRIPTION=Set VAPP_SESSION_LEVEL1 to a value you control before victim login, then call the profile API with the same cookie value after login.
+SESSION_MANAGEMENT_LEVEL1_PAYLOAD_VALUE=document.cookie = "VAPP_SESSION_LEVEL1=your_custom_value"
+SESSION_MANAGEMENT_LEVEL2_CHALLENGE=Use your own login to predict another user's session token. The application generates session IDs from a simple increasing counter.
```
Comment (preetkaran20) [inline src/main/resources/i18n/messages.properties]: I think challenge should be just `Use your own login to predict another user's session token.` and hint should tell `The application generates session IDs from a simple increasing counter.`

File: src/main/resources/i18n/messages.properties [hunk 11]
```diff
@@ -472,6 +472,52 @@ AUTHENTICATION_VULNERABILITY_LEVEL_10_HINT_3=Because the cost is only 4, a commo
...
+SESSION_MANAGEMENT_LEVEL3_CHALLENGE=Decode your session token, find the predictable pattern, and use it to guess another user's encoded session token.
+SESSION_MANAGEMENT_LEVEL3_HINT_1=Encoding does not make a predictable token secure; it only hides the original value.
+SESSION_MANAGEMENT_LEVEL3_HINT_2=Decode the VAPP_SESSION_LEVEL3 cookie and look for the SESSION counter behind it.
+SESSION_MANAGEMENT_LEVEL3_HINT_3=Predict a nearby counter value, encode it the same way, and manually call the profile API using curl, Postman, or browser dev tools with that token.
+SESSION_MANAGEMENT_LEVEL3_PAYLOAD_DESCRIPTION=Decode the issued token, predict a nearby SESSION counter value, encode it again, and call the profile API with the guessed encoded token.
+SESSION_MANAGEMENT_LEVEL3_PAYLOAD_VALUE=Observed token U0VTU0lPTi0xOTAyMA==(SESSION-19020), then try an encoded nearby token such as SESSION-19120.
+SESSION_MANAGEMENT_LEVEL4_CHALLENGE=Access the profile after logout by reusing an old session token. The browser cookie is cleared, but the server-side session remains valid.
```
Comment (preetkaran20) [inline src/main/resources/i18n/messages.properties]: `The browser cookie is cleared, but the server-side session remains valid.` i think this should be a hint not a challenge

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        if (LOGIN_ACTION.equals(action)) {
+            return sessionManagementService.level2Login(username, password);
+        }
+        if (PROFILE_ACTION.equals(action)) {
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: shall we make switch case instead?

**Review by preetkaran20 — CHANGES_REQUESTED** [review]
@CHI7440 The PR looks good to me. Just a few comments overall:
1. Can you please change GET to POST call and use body to pass usernames and passwords? GET is insecure as urls are logged in access.logs as well as browser history.
2. Can you please change token append to url? tokens should be automatically attached by browser as they are cookies
3. Add the string constants so that typo mistakes and future maintenance is easier
4. please fix the concurrency issues as reported by coderabbit.
5. Some cosmetic changes to challenge text
6. Add firstname and lastname to the logged in user so profile returns it and that looks little more realistic.


Overall PR is very close to merging it.

File: src/main/resources/scripts/SessionManagement/db/schema.sql [hunk 13]
```diff
@@ -0,0 +1,11 @@
+DROP TABLE IF EXISTS session_users;
+
+CREATE TABLE session_users (
+    id INT PRIMARY KEY,
+    username VARCHAR(50) NOT NULL,
```
Comment (preetkaran20) [inline src/main/resources/scripts/SessionManagement/db/schema.sql:7]: can you please add firstname and lastname too in this as that makes profile attributes as little more.

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java (hunk 7 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: @preetkaran20, to move login/logout parameters into the request body, those actions should be handled through POST. Since profile is naturally a GET operation, we would need separate mappings per level: one POST mapping for login/logout and one GET mapping for profile. The current implementation keeps login/profile/logout behind a single action-based level endpoint, so I will need to check whether the VulnerableApp endpoint discovery supports multiple mappings for the same level. Can you please confirm on this

Comment (preetkaran20) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: Great point. As this is springboot application so you can create profile endpoint via request mapping annotation. For each level profile endpoint would look something like /LEVEL_1/profile.

in UI you can get the url for a vulnerability level and append profile and call the backend with that.

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java (hunk 2 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java]: Ok got it,  I’ve moved the repeated response keys into constants to keep the code consistent and reduce duplication.

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java (hunk 3 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementService.java]: Done

File: src/main/resources/static/templates/SessionManagementVulnerability/LEVEL_6/SecureSessionManagement.js (hunk 5 above)
Comment (CHI7440) [reply src/main/resources/static/templates/SessionManagementVulnerability/LEVEL_6/SecureSessionManagement.js]: moved username and password into the request body so they are not exposed in the URL or access logs.

File: src/main/resources/static/templates/SessionManagementVulnerability/Common/SessionManagement.js (hunk 6 above)
Comment (CHI7440) [reply src/main/resources/static/templates/SessionManagementVulnerability/Common/SessionManagement.js]: Done

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java (hunk 7 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: Thanks, implemented

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java (hunk 8 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: Done

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java (hunk 9 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java]: Ok thanks, updated

File: src/main/resources/i18n/messages.properties (hunk 10 above)
Comment (CHI7440) [reply src/main/resources/i18n/messages.properties]: Done

File: src/main/resources/i18n/messages.properties (hunk 11 above)
Comment (CHI7440) [reply src/main/resources/i18n/messages.properties]: Done

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java (hunk 12 above)
Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionManagementVulnerability.java]: Done

File: src/main/resources/scripts/SessionManagement/db/schema.sql (hunk 13 above)
Comment (CHI7440) [reply src/main/resources/scripts/SessionManagement/db/schema.sql:7]: Done

**Comment by CHI7440** [comment]
@preetkaran20, Agreed on all the points and implemented the following: 
- Updated the session flow to use cookies and request bodies where needed.
- Split profile into a separate endpoint for each level.
- Added the missing profile fields.
- Kept the challenge wording more user-focused.
- Made the Level 6 lockout atomic and kept the rest lean.

File: src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java [hunk 14]
```diff
@@ -1,28 +1,29 @@
...
+}
+// public class SessionUserRepository {
 
-    private final JdbcTemplate jdbcTemplate;
+//     private final JdbcTemplate jdbcTemplate;
 
-    public SessionUserRepository(@Qualifier("applicationJdbcTemplate") JdbcTemplate jdbcTemplate) {
-        this.jdbcTemplate = jdbcTemplate;
-    }
+//     public SessionUserRepository(@Qualifier("applicationJdbcTemplate") JdbcTemplate jdbcTemplate)
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java]: Please remove commented code

Comment (CHI7440) [reply src/main/java/org/sasanlabs/service/vulnerability/sessionManagement/SessionUserRepository.java]: Done

**Comment by preetkaran20** [comment]
@CHI7440 Thanks a lot for the PR. It looks good to me. Merging it now.

**Comment by preetkaran20** [comment]
🎉 Thanks for contributing @CHI7440!
  We’d love to stay in touch and grow SasanLabs 🚀
  👉 Please fill this (takes 30 sec):  
   https://docs.google.com/forms/d/e/1FAIpQLSfwWVdnULUhtfruA-DN328NwKnBGebaWg9U5y0xivLLxxoMog/viewform?usp=pp_url&entry.1414771947=CHI7440
   
   Also consider ⭐ starring the repo if you like it! 

🎯 **Want to keep going?** Here's a good next issue for you:
👉 [#560 - Add Challenge mode to Path Traversal Vulnerability](https://github.com/SasanLabs/VulnerableApp/issues/560)

**Comment by preetkaran20** [comment]
@all-contributors please add @CHI7440 for code
